src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,139 @@
+            const clause = statement.caseBlock.clauses[i];
+            const previousClause: ts.CaseOrDefaultClause | undefined = statement.caseBlock.clauses[i - 1];
 
-        // If the clause condition holds, go to the correct label
-        const condition = lua.createBinaryExpression(
-            switchVariable,
-            context.transformExpression(clause.expression),
-            lua.SyntaxKind.EqualityOperator
-        );
+            // Skip redundant default clauses, will be handled in final default case
+            if (i === 0 && ts.isDefaultClause(clause)) continue;
+            if (ts.isDefaultClause(clause) && previousClause && containsBreakOrReturn([...previousClause.statements])) {
+                continue;
+            }
 
-        const goto = lua.createGotoStatement(`${switchName}_case_${index}`);
-        return lua.createIfStatement(condition, lua.createBlock([goto]), previousCondition);
-    }, undefined as lua.IfStatement | undefined);
+            // Compute the condition for the if statement
+            if (!ts.isDefaultClause(clause)) {
+                if (condition) {
+                    // Coalesce skipped statements
+                    condition = lua.createBinaryExpression(
+                        condition,
+                        lua.createBinaryExpression(
+                            switchVariable,
+                            context.transformExpression(clause.expression),
+                            lua.SyntaxKind.EqualityOperator
+                        ),
+                        lua.SyntaxKind.OrOperator
+                    );
+                } else {
+                    // Next condition
+                    condition = lua.createBinaryExpression(
+                        switchVariable,
+                        context.transformExpression(clause.expression),
+                        lua.SyntaxKind.EqualityOperator
+                    );
+                }
 
-    if (concatenatedIf) {
-        statements.push(concatenatedIf);
-    }
+                // Skip empty clauses
+                if (clause.statements.length === 0) continue;
 
-    const hasDefaultCase = statement.caseBlock.clauses.some(ts.isDefaultClause);
-    statements.push(lua.createGotoStatement(`${switchName}_${hasDefaultCase ? "case_default" : "end"}`));
+                // Declare or assign condition variable
+                statements.push(
+                    isInitialCondition
+                        ? lua.createVariableDeclarationStatement(conditionVariable, condition)
+                        : lua.createAssignmentStatement(
+                              conditionVariable,
+                              lua.createBinaryExpression(conditionVariable, condition, lua.SyntaxKind.OrOperator)
+                          )
+                );
+                isInitialCondition = false;
+            } else {
+                assert(!isInitialCondition, "Default clause should never be the initial condition");
```

Comment:
True. Initially, the assert was here to prove to me that the code was correct and this was not happening. Maybe just removing the block entirely?

How would I go about testing a situation with two defaults to ensure something sane happens.

Reply:
Just removing it would be fine, it might produce some weird lua but we don't guarantee working lua if you have broken TS anyway. I don't think there is really a nice way to check this for right now, and tbh I don't really care about this situation either, as long as we don't just throw and exit.